File: smilemenu/backend/desktop_entry.py
from pathlib import Path

from .item import LauncherItem


class DesktopEntry:

    @classmethod
    def from_file(cls, path: Path):

        data = {}

        try:
            with path.open(
                "r",
                encoding="utf-8",
                errors="ignore"
            ) as file:

                in_entry = False

                for line in file:
                    line = line.strip()

                    if line == "[Desktop Entry]":
                        in_entry = True
                        continue

                    if line.startswith("["):
                        in_entry = False
                        continue

                    if not in_entry:
                        continue

                    if "=" not in line:
                        continue

                    key, value = line.split(
                        "=",
                        1
                    )

                    data[key] = value

        except Exception:
            return None

        if not data:
            return None

        if data.get(
            "Type",
            "Application"
        ) != "Application":
            return None

        if data.get(
            "Hidden",
            "false"
        ).lower() == "true":
            return None

        if data.get(
            "NoDisplay",
            "false"
        ).lower() == "true":
            return None

        name = data.get("Name")
        command = data.get("Exec")
        icon = data.get("Icon", "")
        
        categories_raw = data.get("Categories", "")
        categories = []
        if categories_raw:
            categories = [
                cat.strip() 
                for cat in categories_raw.split(";") 
                if cat.strip()
            ]

        if not name or not command:
            return None

        return LauncherItem(
            name=name,
            command=command,
            icon=icon,
            categories=categories
        )

File: src/desktopentry.cpp
#include "desktopentry.h"
#include <QFile>
#include <QTextStream>
#include <QDebug>
#include <QStandardPaths>
#include <QLocale>

AppItem* DesktopEntry::fromFile(const QString &path)
{
    QFile file(path);
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
        return nullptr;

    QTextStream in(&file);
    QString line;
    bool inEntry = false;
    QMap<QString, QString> data;
    while (in.readLineInto(&line)) {
        line = line.trimmed();
        if (line == "[Desktop Entry]") {
            inEntry = true;
            continue;
        }
        if (line.startsWith('[')) {
            inEntry = false;
            continue;
        }
        if (!inEntry)
            continue;
        int sep = line.indexOf('=');
        if (sep == -1)
            continue;
        QString key = line.left(sep).trimmed();
        QString value = line.mid(sep + 1).trimmed();
        data[key] = value;
    }

    if (data.isEmpty())
        return nullptr;

    if (data.value("Type") != "Application")
        return nullptr;
    if (data.value("Hidden").toLower() == "true")
        return nullptr;
    if (data.value("NoDisplay").toLower() == "true")
        return nullptr;

    const QString tryExec = data.value("TryExec").trimmed();
    if (!tryExec.isEmpty() && QStandardPaths::findExecutable(tryExec).isEmpty())
        return nullptr;

    QString name = data.value("Name");
    const QString locale = QLocale::system().name();
    const QString language = locale.section('_', 0, 0);
    if (data.contains("Name[" + locale + "]"))
        name = data.value("Name[" + locale + "]");
    else if (data.contains("Name[" + language + "]"))
        name = data.value("Name[" + language + "]");
    QString command = data.value("Exec");
    QString icon = data.value("Icon");
    QStringList categories = data.value("Categories").split(';', Qt::SkipEmptyParts);

    if (name.isEmpty() || command.isEmpty())
        return nullptr;

    return new AppItem(name, command, icon, QString(), categories);
}
